ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java

```diff
@@ -61,6 +68,8 @@ public DistributingExecutor(
         .apply(executionContext, serviceContext)
         .inject(statement);
 
+    checkExecutionPermissions(serviceContext, executionContext, injected.getStatement());
```

Comment:
Why are we moving this check from the `RequestValidator` to the executor? It feels like `RequestValidator` is the right place for it, since it's validating user permissions.

Reply:
I forgot to add a comment in the PR to explain this. But the reason is that performing the permission check for KSQL in a CSAS fails because KSQL does not have DESCRIBE permissions on the sink topic.

The RequestValidator creates the CSAS sink topic just before checking the permissions. This sink topic is created by the Client user. The creation, though, is sandboxed. This means is just a fake topic creation that gives WRITE/READ permissions to the Client user. But, it does not for KSQL, so KSQL fails with authorization denied. Even giving permissions in Kafka for KSQL fails in that part of the code.

The best place I thought was to find the non-fake injector that creates the topic in Kafka. I found it here, in the DistributingExecutor, just before persisting the command. So, now it made sense to me that the permissions checks for the Client and KSQL users are performed before persisting the command and after the User has created the sink topics. The checks should be done on non-sandboxed environments.

Reply:
Is it always the case that if the (client) user creates a sink topic via CSAS, then the KSQL user will have permissions for the topic as well? If so, we can update the sandbox to not only give READ/WRITE permissions to the client user for the sink topic, but also to the KSQL user (within the sandbox).

Or if the KSQL server user does not always automatically get permissions for a sink topic when it is created by the client user, what determines when the KSQL server user does or doesn't get permissions?

Reply:
It's the 2nd case `the KSQL server user does not always automatically get permissions for a sink topic when it is created by the client user`. 

In 5.3 and next 5.4 releases, it is up to the client user (or an Admin) to also grant permissions to the KSQL server user to access the sink topic. KSQL does not have the ability to grant permissions to itself yet. These permissions checks for the KSQL server will give more feedback to the client user about why the command execution failed.

Reply:
> In 5.3 and next 5.4 releases, it is up to the client user (or an Admin) to also grant permissions to the KSQL server user to access the sink topic.

If this is the case, then can't we check whether the KSQL server user has permissions to access the sink topic in the `RequestValidator`? If the topic does not currently exist, then the KSQL server user will not have permissions to access it once it is created. If the topic does currently exist, then the `SandboxedKafkaTopicClient` can delegate the verification to the actual `SandboxedKafkaTopicClient`.

Reply:
There's also another problem with the sandboxed context the `RequestValidator`. If a user has only `create` permissions on the sink topic, then the permission check should fail stating the user does not have `write` permission on the topic just created. Because the sandboxed client gives `read/write` after creation, then we never know if the user does not have the proper permissions to write on the sink topic. If the KSQL server has these permissions, then the CSAS won't fail if the Client user does not have Write.

And if I remove the default permissions from the Sandboxed topic client, then several unit tests fail. I don't recall why I couldn't solve it in a different way, but I had to add the default permissions at that time. Now I found this bug while working on this PR.

I think this check is good to have in the `DistributingExecutor` because it is doing checks on real topics, and not in a sandboxed topic which I don't rely on because the topics used can be real or fake.

Reply:
> Because the sandboxed client gives `read/write` after creation, then we never know if the user does not have the proper permissions to write on the sink topic.

Are you saying that in reality when a user creates a topic, they are not automatically granted READ and WRITE permissions on that topic? If so, the `SandboxedKafkaTopicClient` shouldn't automatically grant these permissions either, right?

Reply:
Right. In reality the user does not get default permissions. For now, the default permissions that the `SandboxedKafkaTopicClient` gets do not conflict anywhere. But I will remove them in a follow-up PR to keep it consistent with the reality. I added these default permissions in the past when working on the `TopicAccessValidator`, so now that it is moved, then these default permissions are not needed.

Reply:
If we remove these default permissions from the `SandboxedKafkaTopicClient`, then we'll be able to perform topic access checks in the `RequestValidator` rather than the executor, right? 

Besides the philosophical argument that the checks should be in the validator because they're performing validation of the command, the other reason I don't think it makes sense to have the checks in the executor is because the executor creates the sink topic before performing the check, but if the check fails we don't want to create the sink topic at all (as in, we should reject the entire command).

Your suggestion of having a separate PR to remove the default permissions in `SandboxedKafkaTopicClient#createTopic()` makes sense to me. Perhaps we can merge that PR first and then come back to this one?

Reply:
Aaah, you're right on that problem. If the sink topic is created by the user, but then the KSQL user does not have permissions, then CSAS will fail, but the sink will be already created.

Let me see if I can remove these default permissions.

Reply:
Just did a quick test without the default permissions, and now I recalled what the problem was.

1. If the user does not have permissions to `Create` the sink topic, the `SandboxedKafkaTopicClient` still allows the user to create the fake topic during the `RequestValidator` and `TopicCreateInjector`.
2. Then, the `TopicAccessValidator` fails with not having `Write` permissions on the sink topic ( (a topic that does not exist yet on the real Kafka client).
3. If I give `Write` permissions on the sink topic, the `RequestValidator` won't fail at all, but the user will fail with not having `Create` permissions now. This during the `DistributingExecutor` which it attempts to create the topic on the real Kafka topic client.

The problem I had before is that the `TopicAccessValidator` cannot check if a user has `Create` permissions on a topic that does not exist, because I need the `describeTopic` call to get the authorized operations. I could check on the Cluster permissions (no the topic), but Kafka can give prefixed permissions on non-existent topics, so that created a problem I couldn't solve, except faking the `SandboxedKafkaTopicClient` to let the user continue, and rely on the `DistributingExecutor` injector to check for user permissions when creating the topic.

But that brought the issue that now KSQL does not check if the user has `Write` permissions on the just created topic.

Reply:
Actually, the #3 is invalid. If ai give `Create/Write` permissions on the sink topic, the `TopicAccessValidator` still fails with not having `Write` permissions because it is checking against the fake Sandboxed topic.

Reply:
> The problem I had before is that the `TopicAccessValidator` cannot check if a user has `Create` permissions on a topic that does not exist, because I need the `describeTopic` call to get the authorized operations.

The `SandboxedKafkaTopicClient` can issue this `describeTopic` call, can it not?  https://github.com/confluentinc/ksql/blob/be2f43d7e1e8d553abc6a4d44d317514e2065d7d/ksql-engine/src/main/java/io/confluent/ksql/services/SandboxedKafkaTopicClient.java#L134

Reply:
What default permissions does the `SandboxedAdminClient` grant?  Just `read` and `write` on created topics? My understanding was that this sandbox should have no real effect, ie. all topic creates and deletes going through it are ignored. So why should the default permissions granted by the client matter?

Reply:
> My understanding was that this sandbox should have no real effect, ie. all topic creates and deletes going through it are ignored. So why should the default permissions granted by the client matter?

The sandboxed clients are used by the request validator to validates requests, i.e., if a statement executes successfully in the sandbox, then the statement passes validation and goes on to be executed. If we have a perfect sandbox, then all requests that succeed in the sandbox will succeed outside the sandbox and vice versa. But if the sandbox is imperfect, e.g., if the sandbox grants `read` and `write` permissions on created topics when this does not happen in reality, then some (CSAS) statements will succeed in the sandbox but fail out the sandbox, which is bad since those statements would pass validation but fail partway through execution.

Ideally, we have a perfect sandbox and continue to validate statements by executing them in the sandbox. However, the tl;dr of my conversation with @spena (which continued offline as well) is that creating this perfect sandbox is difficult as there's no easy interface to check whether a user has `create` permissions for a topic. @spena is following up to see how others perform this check, since it feels it should be a relatively common pattern.

Reply:
Seems we don't have an option to check if a user has CREATE permissions. 

We could use `describeAcls` if we use Kafka ACLs only, but if Kafka uses an external authorization service, like RBAC, then it won't return the permissions.

I synced with other teams, and they depend on Kafka denying or allowing the `createTopic` operation, which is what KSQL is currently doing it. And Kafka does not have plans to integrate the `describeAcls` with external authorization libraries. 

Our only option is to keep this PR as it is. If Kafka allows the user to create the sink topic, but the permission checks fail with KSQL not having WRITE permissions, then the sink topic will be created even if the CSAS command fail.

Reply:
> Seems we don't have an option to check if a user has CREATE permissions.

Then we'll have to do it in the executor by simply trying to create the topic, as you said. @agavra made the suggestion that we have the `KafkaTopicClientImpl` log when it creates topics, so users have a way to confirm whether and when topics were created by KSQL, to help mitigate potential confusion around KSQL creating topics on commands that subsequently fail (e.g., because either the user or KSQL server user does not have WRITE permissions on the newly created topic). 

Regarding the check for whether the user and KSQL server user have WRITE permissions on the sink topic, is it possible to check that piece within the validator, or is it not possible because the topic needs to exist first?

Reply:
It's not possible to do it in the validator. 

Kafka has two ways to check for permissions:
1. With `describeAcls`. This requires the user to have `Describe` permissions on the Cluster. This does not return ACLs handled by an external authorization provider, like RBAC.
2. With `describeTopic`. This returns ACLs and RBAC permissions merged into the returned TopicDescription. However, this requires the topic to exists, so we cannot check for CREATE.

So far, we can only keep the check in the executor, at least for 5.4. If Kafka decides to support RBAC in describeAcls, for instance.